intel_adsp cleanup, pass #2

andyross opened the pull request:
Huge cleanup pass on the cAVS hardware.  In theory no code paths have changed beyond things like operation ordering and minor syntax/API changes, so in theory nothing has broken.  Seems to build for me on all boards and to run everything I've attempted on cavs 1.5 (Apollo Lake Up Squared board) and 2.5 (Tiger Lake chromebook).

It's not quite complete yet.  There is a unification of the linker scripts that I'm still working on.

But the highlight is that this puts essentially all of the board/SoC changes into devicetree and kconfig.  No C code remains in the board and sub-soc directories, just configuration layer files (and the aforementioned linker script).

Commit:
soc: intel_adsp: General soc_mp.c cleanup

Misc cleanup, no non-trivial logic changes.

Swap in new ("rsr <reg>, REGISTER_NAME") syntax for Xtensa SR's in
place of inconsistent usage of the older one ("rsr.REGISTER_NAME
<reg>").

Remove the legacy handling of !KERNEL_COHERENCE cases for allocating
the cpu start record.  That has long been a requirement of
multiprocessor code on this platform.

Remove the synchronous testing of the "alive" flag in
arch_start_cpu().  Nothign about that API is intended to be
synchronous, and in fact the Zephyr SMP layer is already doing the
same trick.

Remove some vestigial dead code at the end of z_mp_entry().  It was
apparently intended to handle the case where a CPU function returned,
but that's not legal anyway.  And it was only enabled in the case
where there was only one CPU anyway, which was an impossible situation
(you can't evercall arch_start_cpu() successfully on a system with
only one core, for obvious reasons -- the only core is already
running!).  Replace with an assertion.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
drivers/ipm: Remove intel_adsp_mailbox driver

This is dead code.  It's based on the cAVS "IPC" mechanism to allow
communication to and from the host CPU.  But there is no test rig in
the Zephyr tree for the protocol defined.  And in fact the only
Zephyr-based user of the IPC mechanism (Sound Open Firmware) has its
own IPC driver and speaks its own protocol with the host kernel.  That
driver needs to migrate into Zephyr soon and this legacy bit is just
confusing.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
soc: intel_asdp: Clean up soc_init() code

Reorganize the initialization code to cleanly separate the platforms
and clarify which code is common.  The #if'ery was sort of a mess.
This is in preparation for an incoming patch that unifies the shim
register definitions across platform variants.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
soc: intel_adsp: Remove old support for cAVS 2.5 "alternate boot"

The linker script (and a little bit of SOF) still has support for an
older mechanism for bootstrapping secondary cores by copying code into
lp-sram from a "manifest" emitted by the linker.  This actually never
worked in Zephyr, and we've implemented a different scheme that uses a
small runtime-copied trampoline instead.

Remove.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
soc: intel_adsp: Clean up shim driver

Each platform was defining its own shim.h header, with slightly
variant field definitions, for a register block that is almost
completely compatible between versions.  This is made worse by the
fact that these represent an API imported fairly early from SOF, the
upstream version of which has since diverged.

Move the existing shim struct into a header ("cavs-shim.h") of its
own, remove a bunch of unused symbols, fill in definitions for some
registers that were left out, correct naming to match the hardware
docs in a few places, make sure all hardware dependencies are source
from devicetree only, and modify existing usage to use the new API
exclusively.

Interestingly this leaves the older shim.h header in place, as it
turns out to contain definitions for a bunch of things that were never
part of the shim register block.  Those will be unified in separate
patches.

Finally: note that the existing IPM_CAVS_IDC driver (soon to be
removed from all the intel_adsp soc's) is still using the old API, so
redeclare the minimal subset that it needs for the benefit of the
platforms in transition.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
soc: intel_adsp: Unify bootloader configuration

The generic bootloader code used a per-device "platform.h" file
imported from SOF.  These turn out to have very little actual content.
Move them to the core directory in a single header for now, pending
some rework to place the settings in devicetree.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
tests/kernel: Add test for CONFIG_SMP_BOOT_DELAY

This feature is in tree and used by the SOF app, but we don't have a
local test for it.  Add one, including a case to track regressions in
a known failure mode (where the second CPU wouldn't get its IDC
interrupts set up correctly if spawned at runtime).

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
drivers: cavs_timer: Use the new interrupt controller API

Recent work to this platform added a new, cleaner low level API to the
interrupt controller.  Replace the hand-cooked register access with
that.  This is still not as good as having proper multicore support in
the intc_cavs driver, but it's at least better.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
soc/intel_adsp: Fix _ASMLANGUAGE declaration for boot_entry.S

The cAVS bootloader code (for... somewhat silly reasons) doesn't build
via the same paths as regular Zephyr object files, so it wasn't
getting the _ASMLANGUAGE define.  That meant that Zephyr headers
defining BIT() were using syntax incompatible with some assemblers
(specifically the Cadence xcc assembly; current gas versions were
fine).

Not 100% sure this is the best spot to put this, but the root fix is
to get the bootloader building into the same link as the rest of
Zephyr anyway.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
soc: intel_adsp: Unify vector table generation

This was an abstraction layer without a purpose.  All existing
platforms have the same (LXn core) layout.  When we need to split this
out in the future, the right thing will be to use the values already
provided by the platform core-isa.h and not duplicate them anyway.

Think of this as a first step to an incoming rework of the Zephyr
Xtensa interrupt entry generation, which is long overdue.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
soc/intel_adsp: Unify HP-SRAM definition

The HP SRAM block address and size is specified in four different ways
(devicetree, "SRAM_*", "HP_SRAM_*" and "L2_SRAM_*" macros).  Unify,
moving the C definition (which just fetches it from dts) to a single
header and out of the platform layer.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
soc/intel_adsp: Move invariant bootloader params out of platform layer

These macros (LPSRAM_MASK, SRAM_BANK_SIZE, HOST_PAGE_SIZE) never
change, and are always used in just one file.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
soc/intel_adsp: Unify fake linker sections

There are multiple "fake" (not part of image) sections needless linked
to explicit addresses right now.  This should be cleaned up, but in
the meantime let's at least put their definitions all in one place so
they aren't cut/pasted into every platform.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
soc/intel_adsp: Unify "RAM" region

There was a divergent set of definitions for a "RAM" region for the
benefit of memory layout in the platform headers.  In fact there was
only one platform dependence (cAVS 1.5 has 32k instead of 64k
reserved).  Put that into kconfig in a single place, and add a warning
that this is a trap region with hidden dependencies in both Zephyr and
SOF.  Good enough until we clean this up and make everything visible
to the linker.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
soc/intel_adsp: Unify "firmware status register" window size

There is a diverged definition for SRAM_REG_FW_END, which exists to
prevent the Zephyr window initialization from changing values that
were already set by the ROM or bootloader (though this is incomplete,
as we're not ensuring the memory is actually the same space except by
convention; we also don't have any Zephyr-side visibility as to the
content of this struct).

That was silly; the only thing worse than one magic number is four
magic numbers in different files.  Write a formula that works for all
the platforms and put it in the C file where it's used.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
soc/intel_adsp: Unify host window register interface

These registers are identical on all platforms, the only difference
being that cAVS 1.5 places them at a different address.

Create a devicetree node to track the register block, and replace the
platform header code with a global API defined once (it works like the
pre-existing shim struct).

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
soc/intel_adsp: Remove magic entry point addresses

The platform layer (except cAVS 1.5, which computed it via alternate
means as an offset on other stuff) was specifying the entry point as
an explicit address needlessly.  In fact the linker scripts already
are written to place the entry point at the first address of linkable
RAM, which is already available as the RAM_BASE symbol.

Unify.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
soc/intel_adsp: Unify LP-SRAM sizing

Same as for the HP-SRAM memory region.  It's already exposed in
devicetree, so take the per-platform values out (including some dead
code on 2.5) and put them in a global header.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
soc/intel_adsp: Clean up host memory window layout

The locations of the memory windows within the reserved SRAM region
was being construted via a bunch of magic numbers referencing
(potentially historical?) SOF usage of those areas.  But Zephyr only
ever touches two of these windows, and only cares about the sizes and
offsets.  The complexity was hurting and not helping (especially since
there was no attempt made to unify these values with the ones that are
actually live in the SOF tree).

Replace with kconfig variables that simply specify the offset.  Only
one platform has a nonstandard layout anyway.  That allows SOF to move
things around in a clean way if it wants.  Ideally we should be
presenting a proper API for managing this region, though.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
soc/intel_adsp: Unify boot loader memory layout

These values (used to arrange the bootloader within IMR memory) are
mostly computed or fixed to values that don't chagne between
platforms.  Only the manifest address and the location of the data
section change.  Put those in kconfig, move the rest to the global
header.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
soc/intel_asdp: Unify misc. initialization registers

These two register blocks are defined in the platform layers, but
never change (except on 1.5 where they don't exist).  I don't want to
write a full devicetree interface for them as I can't find good docs
currently.  They are used only at system initialization, so move the
definitions to the single file where they're used.  In the longer term
we will want to move at least the GPDMA setup into a driver anyway.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
soc/intel_adsp: Unify/dtsify L2 local memory control block

These registers were hardwired in the platform layer.  Move to
devicetree, via a struct interface that looks like the pre-existing
shim layer.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
soc/intel_adsp: Eliminate platform configuration headers

All the in-use contents of these files have now been moved to the
intel_adsp core, and they are configured via devicetree and kconfig.
Remove the legacy headers.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
soc/intel_adsp: Move generated interrupt routines to common code

While technically this file can change for any instantiation of Xtensa
hardware, in practice all these devices have identical interrupts
setups.  These files were duplicates, so there's no value in keeping
them in per-sub-soc directories.  (Really we should wire it up so that
the generator gets run automatically with the build, but that will
need to wait for a rework of interrupt entry).

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
soc/intel_adsp: Unify linker scripts

While the linker scripts for these platforms had diverged in form, the
behavior remained compatible.  Link all cAVS devices with the same
linker script included from the common directory (it's a verbatim copy
of the cavs_v25 script).

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

kv2019i reviewed (changes requested):
Breaks SOF build to:
```
In file included from /srv/home/jenkins/workspace/sof_generic_build/zephyrproject/zephyr/include/devicetree.h:22, from /srv/home/jenkins/workspace/sof_generic_build/zephyrproject/zephyr/soc/xtensa/intel_adsp/common/include/cavs-mem.h:7, from /srv/home/jenkins/workspace/sof_generic_build/zephyrproject/zephyr/soc/xtensa/intel_adsp/common/bootloader/base_module.c:11: /srv/home/jenkins/workspace/sof_generic_build/zephyrproject/zephyr/include/sys/util.h:48:3: error: #error Missing required predefined macros for BITS_PER_LONG calculation [13/253] Generating misc/generated/syscalls.json, misc/generated/struct_tags.json
```

lyakh reviewed (changes requested):
A lot of good changes, thanks! But this should really be split in multiple PRs, and some commits should be split too IMHO, as confirmed, e.g. by a failing multi-core boot on APL.

andyross commented:
Repush, with various patches (some new) split out in #40345 #40346 and #40347 (should resolve most of the comments about the 1.5 IDC driver changes, I think).

Note that a xcc build seems to be working for me (albeit with a different toolchain version in a slightly different tree).  I saw a similar-looking fix go by recently, so it's possible this resolved itself?

The series here is harder to split, since it's all just steady progress at unification.  I can work at that too if folks have trouble following, but honestly it seems OK to me.

andyross commented:
Sorry, I tested stuff yesterday evening in the wrong order: there **remains** a hard ordering dependency on #40345 here.  The old ipm_idc driver is (obviously) still written to the old register API, which is removed in this PR.  That driver as a whole was removed in the IDC pull request, so it's not a problem in combination.  But there's zero point to trying to port it here just to remove it there.  Not a problem as long as we get things merged promptly.

(There's a much simpler dependency I missed on #40346 too.  But that is just a quick timer driver fix so I'm sure it'll be in  soon.)

kv2019i commented:
Hmm, I'm still getting build errors 4f636ce88fdf74f55976060e02115d2d4f119813  when I build SOF with sof/scripts/xtensa-build-zephyr.sh 

gcc:
/mnt/sof/zephyrproject/zephyr/drivers/timer/cavs_timer.c: In function 'count64':
/mnt/sof/zephyrproject/zephyr/drivers/timer/cavs_timer.c:80:17: error: 'shim_regs' undeclared (first use in this function)
   80 |  uint64_t ret = shim_regs->walclk32_lo;

xcc:
In file included from /mnt/sof/zephyrproject/zephyr/include/devicetree.h:22,
                 from /mnt/sof/zephyrproject/zephyr/soc/xtensa/intel_adsp/common/include/cavs-mem.h:7,
                 from /mnt/sof/zephyrproject/zephyr/soc/xtensa/intel_adsp/common/bootloader/boot_module.c:11:
/mnt/sof/zephyrproject/zephyr/include/sys/util.h:48:3: error: #error Missing required predefined macros for BITS_PER_LONG calculation

Commit:
soc/intel_adsp: Add workarounds for building bootloader with xcc

The bootloader code on this SOC has its own cmake rules, which means
it doesn't get the Zephyr-specific magic from the toolchain layer and
the code needs to handle fixups manually.  (Specifically: we have a
xcc_missing_defs.h header to provide gcc symbols that xcc doesn't
have, and assembly needs to be built with _ASMLANGUAGE so headers
don't include C syntax.)

Long term the right solution here is to build the bootloader as part
of the Zephyr binary.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andyross commented:
Think I found the xcc interaction.  The cavs_timer issue just required the separate (now-merged-and-rebased) timer fix.  And again, building intel_adsp_cavs15 still requires the separate IDC change.

kv2019i commented:
@andyross wrote:
> Think I found the xcc interaction. The cavs_timer issue just required the separate (now-merged-and-rebased) timer fix. And again, building intel_adsp_cavs15 still requires the separate IDC change.

Ok, builds now with both gcc and xcc (latest version of this PR + https://github.com/zephyrproject-rtos/zephyr/pull/40345 ).

andyross commented:
Rebase to fix a collision in the linker scripts vs. #40319

andyross commented on soc/xtensa/intel_adsp/cavs_v18/include/soc/shim.h:
```diff
@@ -213,21 +24,10 @@
-#define SHIM_GPDMA_CHLLPC_EN		BIT(7)
-#define SHIM_GPDMA_CHLLPC_DHRS(x)	SET_BITS(6, 0, x)
-
-#define L2LMCAP			0x71D00
```
The L2LM block is a separate register block (indeed it only exists on 1.8+, used for memory PM control which we don't do in 1.5).  It becomes a "cavs-l2lm" devicetree block and is available at runtime as a volatile struct pointer as CAVS_L2LM

andyross commented on soc/xtensa/intel_adsp/common/adsp.c:
```diff
@@ -31,9 +31,9 @@ LOG_MODULE_REGISTER(sof);
-		    DMWBA(0));
+	CAVS_WIN[0].dmwlo = HP_SRAM_WIN0_SIZE | 0x7;
+	CAVS_WIN[0].dmwba = (HP_SRAM_WIN0_BASE | CAVS_DMWBA_READONLY
+			     | CAVS_DMWBA_ENABLE);
```
It's not clear in the github UI, but the parentheses are an easy way to get editors to align multi-line expressions like this.  I do this all over the place in Zephyr; no idea if there's a coding style statement anywhere, but it's pretty common.

andyross commented on soc/xtensa/intel_adsp/common/bootloader/boot_entry.S:
```diff
@@ -16,10 +16,12 @@
 #include <xtensa/config/core-isa.h>
+#include "platform.h"
+
+#define PLATFORM_PRIMARY_CORE_ID 0
```
I don't see a mailbox.h in the tree?  Maybe that got removed somewhere else, or do you mean a SOF header?  Actually one of the side benefits to removing the old API headers is that we avoid tripping over spots where SOF was pulling in Zephyr symbols instead of using its own or moving to the Zephyr API (in this case the SMP CPU struct).

andyross commented on soc/xtensa/intel_adsp/common/bootloader/boot_entry.S:
```diff
@@ -66,7 +72,7 @@ wnd0_error_address:
 #if defined(PLATFORM_MEM_INIT_AT_BOOT)
 shim_ldoctl_address:
-	.word SHIM_BASE + SHIM_LDOCTL
+	.word (SHIM_ADDR + 0xa4)
```
It's a struct offset.  It's not better, though in this case it's used in only one space, as a variant (assembly) API to something we're doing in C elsewhere.  Actually in a future PR, virtually all of the boot loader entry assembly is disappearing and I'll be doing everything in C.  But I can certainly fix this here.

andyross replied:
Actually I take that back.  The symbol name in the line immediately above identifies this immediate constant as the "shim_ldoctl_address".  Adding a SHIM_LDOCTL_OFFSET symbol would add no clarity, IMHO.  I think this is fine, personally.  Again, the ultimate goal is not to be using these registers in assembly at all.

andyross commented on soc/xtensa/intel_adsp/common/include/cavs-shim.h:
```diff
@@ -0,0 +1,112 @@
+	};
+	uint32_t _unused5[4];
+	uint32_t l2mecs;
+	uint32_t l2mpat;
```
Same deal, what really happened is that they moved out the L2LM registers to be their own register block.

andyross commented on soc/xtensa/intel_adsp/common/include/cavs-shim.h:
```diff
@@ -0,0 +1,112 @@
+};
+#endif /* _ASMLANGUAGE */
+
+#define CAVS_SHIM (*((volatile struct cavs_shim *)DT_REG_ADDR(DT_NODELABEL(shim))))
```
That's exactly what this is.  :) The older definition went away and the newer one moved into a header and fetches the address from devicetree.

andyross commented on soc/xtensa/intel_adsp/common/soc.c:
```diff
@@ -210,60 +206,43 @@ static void power_init_v15(void)
-		SHIM_CLKCTL_TCPLCG_DIS_ALL;
+	CAVS_SHIM.clkctl = (CAVS_CLKCTL_RHROSCC |
+			    CAVS_CLKCTL_OCS |
+			    CAVS_CLKCTL_LMCS);
```
These are the same value.  The older macros were mostly zeroes.  Which would sort of be OK from a documentation standpoint, except we didn't have macros defining all the fields in CLKCTL, so this was sort of just a random assortment anyway.  I cut bait and left the zero-valued defaults unstated and only define the values we're changing.  A better API here would be a macro taking values for all the fields in the register, I guess.  But then that would end up exposing a lot of detail which is poorly documented even in internal docs.

andyross replied:
Note that the comments on the field values live in cavs-shim.h now (c.f. CAVS_CLKCTL_* macros).   But IMHO there's only so much that can be done in source code -- we aren't hardware docs (and frankly the hardware docs aren't as much help as they should be anyway).

andyross commented on soc/xtensa/intel_adsp/common/soc.c:
```diff
@@ -281,13 +275,17 @@ static int soc_init(const struct device *dev)
+	}
 
-#if CONFIG_MP_NUM_CPUS > 1
 	soc_idc_init();
```
Because the hardware is present either way and still needs to be initialized.  If you look at the implementation of soc_idc_init() it's just enumerating over CONFIG_MP_NUM_CPUS anyway, so there's minimal wasted effort.

andyross commented on soc/xtensa/intel_adsp/common/soc_mp.c:
```diff
@@ -66,6 +66,15 @@ static struct k_spinlock mplock;
+ * it happens, SOF still emits the code to copy this data, so it needs
+ * to see this symbol point to a zero.
+ */
+uint32_t _loader_storage_manifest_start;
```
Mostly because SOF internals are scarier and I wasn't sure what kind of thing that would break.  We don't want SOF to use the old-style IDC message code either, we just want it to do nothing.

andyross commented on soc/xtensa/intel_adsp/common/soc_mp.c:
```diff
@@ -251,9 +250,24 @@ void z_mp_entry(void)
 
-#ifdef CONFIG_SMP_BOOT_DELAY
-	cavs_idc_smp_init(NULL);
-#endif
```
That's handled in soc_idc_init() now.

lgirdwood replied:
I assume the plan is to move away from IPM for all CAVS platforms since IIRC it only supports 2 cores ?

andyross commented on soc/xtensa/intel_adsp/common/soc_mp.c:
```diff
@@ -354,27 +368,18 @@ void arch_start_cpu(int cpu_num, k_thread_stack_t *stack, int sz,
-	k_busy_wait(100);
-
-#ifdef CONFIG_SMP_BOOT_DELAY
-	cavs_idc_smp_init(NULL);
```
IDC is a global device, it's initialized already.

andyross commented on tests/kernel/smp_boot_delay/src/main.c:
```diff
@@ -0,0 +1,84 @@
+volatile bool mp_flag;
+
+struct k_thread cpu1_thr;
+K_THREAD_STACK_DEFINE(thr_stack, 2048);
```
Good catch, will fix.

andyross commented on tests/kernel/smp_boot_delay/testcase.yaml:
```diff
@@ -0,0 +1,4 @@
+tests:
+  kernel.multiprocessing.smp_boot_delay:
+    tags: kernel smp
+    platform_allow: intel_adsp_cavs15 intel_adsp_cavs18 intel_adsp_cavs25
```
In theory intel_adsp_cavs20 is unified in the same way everything else is (AFAICT it's 100% indistinguishable from 1.8 -- literally the only change I could find is a clock rate).   In practice no one has such hardware.  I guess my intuition is that we should enable it only when we know we care about it, but I can add it here if you really want it.

andyross commented:
Rebase around a minor collision in CODEOWNERS.  Also add some comments in cavs-shim.h regarding functionality that got moved to the L2LM block in 1.8+, fix the 2048-vs-STACKSZ goof in the boot delay tests, and make that test run on cavs20.

lgirdwood commented on soc/xtensa/intel_adsp/common/soc.c:
```diff
@@ -210,60 +206,43 @@ static void power_init_v15(void)
+			    CAVS_CLKCTL_OCS |
+			    CAVS_CLKCTL_LMCS);
 
+#ifndef CONFIG_SOC_SERIES_INTEL_CAVS_V15
```
Do we know if SOF is doing the same for v1.5 ?

lgirdwood reviewed (approved):
One open on the APL DMA PM gating, but we can resolve this incrementally if needed.

Commit:
soc/intel_adsp: Add support for legacy IDC driver for pre-2.5 hardware

The IDC driver was written for Tiger Lake era devices, but works fine
on the earlier hardware too.  Make it selectable; if you don't
configure IPM_CAVS_IDC, then you get the new driver.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andyross commented:
I added a series to the end that "retro-ports" the older API such that the IPM driver will continue to build in a kconfig-selectable tandem with the new one.  This hopefully allows this PR to merge in isolation without breaking anything.

It then breaks up the controversial patch from the (now-closed) #40345 and includes the bits here enable (but don't configure by default) the new driver for cavs15/18/20.

See the branch https://github.com/andyross/zephyr/tree/cavs-cleanup-tmp for a continuation of the series that disables the older IPM driver for each of the platforms that used it and eventually removes support entirely.  Presumably that still fails on 1.5, but please test?

kv2019i reviewed (approved):
Reviewed and tested with SOF in following configs:

- cavs15 single-core (nocodec tplg)
- cavs15 multi-core (nocodec tplg)
- cavs25 single core (nocodec tplg)
- SKIPPED cavs25 multicore (known bug https://github.com/thesofproject/sof/issues/4731 )

@marc-hb  to add test results for cavs18

lyakh reviewed (commented):
I like this! Thanks a lot for this huge and much needed clean up! My only remaining (significant AFAICS) concern is bisection breakage. These are 29 patches, at least some with non-trivial changes, so the chance of a regression isn't negligible I'd say. And so this must be bisectable... Other comments are mostly minor remarks or even questions, maybe there's a reason for all those changes too.

andyross commented on drivers/ipm/ipm_cavs_idc.h:
```diff
@@ -7,6 +7,26 @@
+#define IPC_IDCIETC_DONE      BIT(30)
+#define IPC_IDCCTL            0x50
+#define IPC_IDCCTL_IDCTBIE(x) BIT(x)
+#endif
```
It is now, yeah.  The original order of the series was that everything was moved to the new API and then the headers were removed.  Now that's at the end, so indeed there's a hole.  Will fix.

andyross commented on drivers/timer/cavs_timer.c:
```diff
@@ -195,9 +196,6 @@ void smp_timer_init(void)
-			+ CAVS_ICTL_INT_CPU_OFFSET(arch_curr_cpu()->id)
-			+ 0x04,
-		    22 + TIMER);
+	CAVS_INTCTRL[arch_curr_cpu()->id].l2.clear = CAVS_L2_DWCT0;
```
It's still correct.  These two bits of code are identical (the new API just names the register being used instead of using an offset).  The complaint here is that the cavs_ictl interrupt controller driver only understands global masking and doesn't hook the per-core mask registers.  So the individual device drivers need to do it manually.  Definitely a spot for cleanup in the future.

andyross commented on soc/xtensa/intel_adsp/common/bootloader/boot_loader.c:
```diff
@@ -6,6 +6,11 @@
+/* Older (GCC 4.2-based) XCC variants need a fixup file. */
+#if defined(__XCC__) && (__GNUC__ == 4)
+#include <toolchain/xcc_missing_defs.h>
+#endif
```
That's a good point.  This was exposed because the bootloader is now including more Zephyr headers than it used to and ran afoul of this warning (for which we had a workaround, but done in a way that only works for "official" Zephyr .obj files and not side-built things like the bootloader).  Will move so it lives before the header changes.

andyross reviewed (commented):
It's true that bisectability got broken.  The original scheme had the old driver removed first, and the changes needed to split it out to a separate PR left a hole.  Will address.

andyross commented on drivers/ipm/ipm_cavs_idc.h:
```diff
@@ -7,6 +7,26 @@
+/* Redeclaration of the earlier IDC register API for platforms being
+ * held back on this driver.
+ */
+#ifndef CONFIG_SOC_SERIES_INTEL_CAVS_V25
```
As of this PR, the old driver is symmetric and "can" be used, so I figured I might as well declare it that way.  The reason for specifying 2.5 specifically is that it could never use IPM (which limits us to 2 cores) and is limited to the new driver only.

Commit:
soc/intel_adsp: Clean up IPM compiler guards

The CAVS_IDC_IPM driver happens to be used only on non-2.5 hardware,
but it's best to be clear in the conditional compilation when we're
talking about hardware-dependencies and when we mean software
configuration.  This was mixed up in a few spots.

Also fix a warning that creeps in on non-default drivers choices about
an undeclared ipm function.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
soc/intel_adsp: Simplify boot ROM protocol

There was an attempt in the old code to express this as a formal
protocol with a proper field definitions, etc...  But in fact no such
protocol really exists.  This scheme is only used in one place to send
one specific message to code fixed in ROM on legacy devices that only
knows how to recognize this specific value.  And 2.5 and later
hardware are moving away from it anyway.

Just express it directly, and explain in comments.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andyross commented:
Rebase (cavs-cleanup-tmp also got the same treatment), slice up a few patches to make things happen in a bisectable sequence.

The one thing I didn't add was the GPDMA clock handling.  I realize now that the reason it wasn't done with the shim block is that it's not a shim thing: it's a different register block entirely and needs its own devicetree entry.  I'll finish that and submit post-PR (or maybe slide it into the cavs-cleanup-tmp for early review).

andyross commented:
Any motion on this one?  @lyakh do the bisection fixups look OK to you?  Would be really nice to get this merged this week.

kv2019i commented:
@andyross wrote:
> Any motion on this one? @lyakh do the bisection fixups look OK to you? Would be really nice to get this merged this week.

W.r.t. bisects, "soc/intel_adsp: Unify HP-SRAM definition"  breaks the APL build. I think you need to move "soc/intel_adsp: Add workarounds for building bootloader with xcc" earlier in the series.

Otherwise bisect works in the series. I did a limit 10 round bisect on cavs15 hw and aside that build error, no issues found. FYI @lyakh

marc-hb commented:
I tested the tip of this PR (0b8a45cb15377ca55) on my WHL UpXtreme and it does not seem less stable than without it.

> do the bisection fixups look OK to you? 

> W.r.t. bisects, "soc/intel_adsp: Unify HP-SRAM definition" breaks the APL build.

It's not just about bisection, it's also about daily tests and test coverage. Right now we have zero "west-like" control on versions. We should get a bit more soon but it will still be limited or manual or both.

> Fun fact: I rarely submit HEAD to github but HEAD~N for some value of N :-)

lgirdwood commented on soc/xtensa/intel_adsp/common/include/cavs-shim.h:
```diff
@@ -0,0 +1,112 @@
+	uint32_t dbgo;
+	uint32_t svcfg;
+	uint32_t _unused9[2];
+};
```
Good point, there are a few uint16_t in there too and we can never tell what a new compiler will do. This could also be aligned onto a uint32_t too.

lgirdwood commented on soc/xtensa/intel_adsp/Kconfig:
```diff
@@ -12,6 +12,15 @@ config SOC_FAMILY
+	  Bytes to reserve at the start of HP-SRAM.  Zephyr will not
+	  place any symbols here, though the host windows have
+	  addresses here.  The SOF application also makes direct use
+	  of this region, so be very careful changing this value.
```
Don't we only have the reset and other vectors here, e.g. irq, window spill for Zephyr alongside the shared windows ?

lgirdwood commented on soc/xtensa/intel_adsp/Kconfig:
```diff
@@ -21,6 +21,22 @@ config HP_SRAM_RESERVE
+	  Offset of the first host shared memory window (used for the
+	  "firmware status register" at boot) within the reserved
+	  region at the start of HP-SRAM.
+
```
is long term thinking to partition the windows and permissions in DT ? Seem like a good place ?

lgirdwood reviewed (approved):
LGTM, on basis of updates currently agreed.

lyakh reviewed (approved):
Thanks again!

andyross commented:
Sneaking some replies in post-merge:

> long term thinking, we could eventually put these in an init ELF section and add them heap if they are used at power on only ?

Yeah.  That's the next new feature on the list after I get the internal tree synchronized with this work, actually.  Putting code in IMR should be as simple as an "void __imr my_func()" definition.

> Don't we only have the reset and other vectors here, e.g. irq, window spill for Zephyr alongside the shared windows ?

No vector code, at least in the Zephyr link, those go at the start of the .text region.  But the offsets are still in the SOF tree and seem to be used I think (maybe), so I put this warning here to scary people away from modifying this.  The existing memory layout is essentially a magic configuration pending a bunch of auditing.

> is long term thinking to partition the windows and permissions in DT ? Seem like a good place ?

Probably?  The only firm requirement is that with window 0 we need to make sure we present a consistent picture to the loader, since the fw_status region is shared between ROM and OS.  Right now that means we have to leave it in place exactly where the ROM put it.  But if we were careful (and the hardware cooperates, I certainly haven't tested!) we ought to be able to move it atomically.

At that point the windows can just living in memory as perfectly normal symbols, you could have, e.g.:

    uint32_t __aligned(64) window0[2048];

And let the linker put it wherever it needs to, with no manual mess and no waste.  Something to investigate for sure.

> I assume the plan is to move away from IPM for all CAVS platforms since IIRC it only supports 2 cores ?

Yes, and it's already submitted in #40609.  It had to be removed here because of the regression on cAVS 1.5, but (fingers crossed) I think we managed to figure out a root cause on that this morning.
